runtime/src/main/java/org/corfudb/runtime/view/stream/AddressMapStreamView.java

```diff
@@ -84,11 +86,16 @@ protected ILogData removeFromQueue(NavigableSet<Long> queue) {
                 // to be serviced immediately, rather than reading one entry at a time.
                 ld = read(currentRead, queue);
 
-                if (queue == getCurrentContext().readQueue && ld != null) {
+                if (queue == getCurrentContext().readQueue) {
                     // Validate that the data entry belongs to this stream, otherwise, skip.
                     // This verification protects from sequencer regression (tokens assigned in an older epoch
                     // that were never written to, and reassigned on a newer epoch)
-                    if (ld.containsStream(this.id) && ld.getType() == DataType.DATA) {
+                    if (ld.getType() == DataType.HOLE || (ld.containsStream(this.id) &&
+                            ld.getType() == DataType.DATA)) {
+                        // Convert hole to NOP
+                        if (ld.isHole()) {
+                            ld = new LogData(DataType.DATA, SMREntry.getNOP(), ld.getMetadataMap());
+                        }
                         addToResolvedQueue(getCurrentContext(), currentRead);
```

Comment:
This layer only understands `ILogData` can you move this conversion to `StreamViewSMRAdapter` which actually maps `ILogData` to `SMREntry`